### PR TITLE
HIP-376: Support for Approve/Allowance and ERC20/ERC721 transferFrom

### PR DESCRIPTION
**Pull request by mustafauzunn**
**Description**:

Extends ERC20 and ERC721 support in HTS precompiled contract added with HIP-218 with support for functions for approve/allowance and transferFrom.

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java [hunk 1]
```diff
@@ -238,7 +253,17 @@ public class HTSPrecompiledContract extends AbstractPrecompiledContract {
 	//transferFrom(address sender, address recipient, uint256 amount)
 	//transferFrom(address from, address to, uint256 tokenId)
 	public static final int ABI_ID_ERC_TRANSFER_FROM = 0x23b872dd;
-
+	//allowance(address token, address owner, address spender)
+	protected static final int ABI_ID_ALLOWANCE = -0x229d12c2;
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java:257]: Perhaps we don't need the `-` sign?

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java [hunk 2]
```diff
@@ -1367,8 +1460,9 @@ public TransactionBody.Builder body(final Bytes input, final UnaryOperator<byte[
 
 			final var nestedInput = input.slice(24);
 			super.transferOp = switch (nestedInput.getInt(0)) {
-				case ABI_ID_ERC_TRANSFER -> decoder.decodeErcTransfer(nestedInput, tokenID, callerAccountID,
-						aliasResolver);
+				case ABI_ID_ERC_TRANSFER -> decoder.decodeErcTransfer(nestedInput, tokenId, callerAccountID, aliasResolver);
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java:1463]: We can rename `decodeErcTransfer` -> `decodeERCTransfer` to be compliant with `decodeERCTransferFrom()`.

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java [hunk 3]
```diff
@@ -1527,6 +1652,278 @@ public Bytes getSuccessResultFor(final ExpirableTxnRecord.Builder childRecord) {
...
+		@Override
+		public Bytes getSuccessResultFor(final ExpirableTxnRecord.Builder childRecord) {
+			final TransactionalLedger<AccountID, AccountProperty, MerkleAccount> accountsLedger = ledgers.accounts();
+			final var allowance = (TreeMap<FcTokenAllowanceId, Long>) accountsLedger.get(allowanceWrapper.owner(), FUNGIBLE_TOKEN_ALLOWANCES);
+			long value = 0;
+			for (Map.Entry<FcTokenAllowanceId, Long> e : allowance.entrySet()) {
+				if (allowanceWrapper.spender().getAccountNum() == e.getKey().getSpenderNum().longValue()) {
+					value = e.getValue();
+				}
+			}
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java:1679]: You can extract the logic below in a default method in the Precompile interface and reuse it where it's needed (e.g. ApprovePrecompile:

```
default long getFungibleAllowanceValueForAccount(final AccountID owner, final long accountNum) {
    final var allowance = (TreeMap<FcTokenAllowanceId, Long>) accountsLedger.get(allowanceWrapper.owner(), 
         FUNGIBLE_TOKEN_ALLOWANCES);
			long value = 0;
			for (Map.Entry<FcTokenAllowanceId, Long> e : allowance.entrySet()) {
				if (allowanceWrapper.spender().getAccountNum() == 
                           e.getKey().getSpenderNum().longValue()) {
					value = e.getValue();
				}
			}
}
```

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java [hunk 4]
```diff
@@ -1527,6 +1652,278 @@ public Bytes getSuccessResultFor(final ExpirableTxnRecord.Builder childRecord) {
...
+			return super.body(input, aliasResolver);
+		}
+
+		@Override
+		public Bytes getSuccessResultFor(final ExpirableTxnRecord.Builder childRecord) {
+			final TransactionalLedger<AccountID, AccountProperty, MerkleAccount> accountsLedger = ledgers.accounts();
+			final var allowance = (TreeMap<FcTokenAllowanceId, Long>) accountsLedger.get(allowanceWrapper.owner(), FUNGIBLE_TOKEN_ALLOWANCES);
+			long value = 0;
+			for (Map.Entry<FcTokenAllowanceId, Long> e : allowance.entrySet()) {
+				if (allowanceWrapper.spender().getAccountNum() == e.getKey().getSpenderNum().longValue()) {
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java:1676]: Shouldn't we also check whether the allowance info is for the same `EntityNum tokenNum` that we had invoked in the `redirect` call?

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java [hunk 5]
```diff
@@ -1527,6 +1652,278 @@ public Bytes getSuccessResultFor(final ExpirableTxnRecord.Builder childRecord) {
...
+					if (!OK.equals(checkResponseCode)) {
+						throw new InvalidTransactionException(checkResponseCode);
+					}
+
+					final var approveAllowanceLogic = approveAllowanceLogicFactory.newApproveAllowanceLogic(
+							accountStore, tokenStore, dynamicProperties);
+					approveAllowanceLogic.approveAllowance(transactionBody.getCryptoApproveAllowance().getCryptoAllowancesList(),
+							transactionBody.getCryptoApproveAllowance().getTokenAllowancesList(),
+							transactionBody.getCryptoApproveAllowance().getNftAllowancesList(),
+							EntityIdUtils.accountIdFromEvmAddress(frame.getSenderAddress()));
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java:1770]: Lines 1755-1770 could be extracted in a method in the ApprovePrecompile. This method. could be used both for ERC20 and ERC721 tokens.

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java [hunk 6]
```diff
@@ -1527,6 +1652,278 @@ public Bytes getSuccessResultFor(final ExpirableTxnRecord.Builder childRecord) {
...
+				} else {
+					final var checkResponseCode = allowanceChecks.allowancesValidation(transactionBody.getCryptoApproveAllowance().getCryptoAllowancesList(),
+							transactionBody.getCryptoApproveAllowance().getTokenAllowancesList(),
+							transactionBody.getCryptoApproveAllowance().getNftAllowancesList(),
+							payerAccount,
+							currentView);
+
+					if (!OK.equals(checkResponseCode)) {
+						throw new InvalidTransactionException(checkResponseCode);
+					}
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java:1763]: We can replace this `if` with `validateTrue(OK.equals(checkResponseCode), checkResponseCode);`

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/Utils.java [hunk 7]
```diff
@@ -28,6 +28,7 @@
 import com.hederahashgraph.api.proto.java.ContractID;
 import com.hederahashgraph.api.proto.java.Key;
 import com.hederahashgraph.api.proto.java.NftTransfer;
+import com.hederahashgraph.api.proto.java.ContractID;
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/Utils.java:31]: Probably unnecessary import?

**Review by tinker-michaelj — APPROVED** [review]
LGTM! 🥂

**Pull request merged**